test/e2e/kubebuildercontext.go

```diff
@@ -0,0 +1,169 @@
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	. "github.com/onsi/ginkgo"
+)
+
+type CmdContext struct {
+	// environment variables in k=v format.
+	Env []string
+	Dir string
+}
+
+// KbContext specified to run e2e tests
+type KbContext struct {
+	*CmdContext
+	Domain    string
+	Group     string
+	Version   string
+	Kind      string
+	Resources string
+	ImageName string
+	Kubectl   *Kubectl
+}
+
+// ContextWithSuffix init with a random suffix for test KbContext stuff,
+// to avoid conflict when running tests synchronously.
+func ContextWithSuffix(testSuffix string, env []string) (*KbContext, error) {
+	testGroup := "bar" + testSuffix
+	path, err := filepath.Abs("e2e-" + testSuffix)
+	if err != nil {
+		return nil, err
+	}
+
+	cc := &CmdContext{
+		Env: env,
+		Dir: path,
+	}
+
+	return &KbContext{
+		Domain:     "example.com" + testSuffix,
+		Group:      testGroup,
+		Version:    "v1alpha1",
+		Kind:       "Foo" + testSuffix,
+		Resources:  "foo" + testSuffix + "s",
+		ImageName:  "e2e-test/controller-manager:" + testSuffix,
+		CmdContext: cc,
+		Kubectl: &Kubectl{
+			Namespace:  fmt.Sprintf("e2e-%s-system", testSuffix),
+			CmdContext: cc,
+		},
+	}, nil
+}
+
+func (cc *CmdContext) RunCommand(cmd *exec.Cmd) ([]byte, error) {
+	cmd.Dir = cc.Dir
+	cmd.Env = append(os.Environ(), cc.Env...)
+	command := strings.Join(cmd.Args, " ")
+	fmt.Fprintf(GinkgoWriter, "running: %s\n", command)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return output, fmt.Errorf("%s failed with error: %s", command, string(output))
+	}
+
+	return output, nil
+}
+
+// Preparewd prepare a work directory for testing
+func (kc *KbContext) Preparewd() error {
+	fmt.Fprintf(GinkgoWriter, "preparing testing directory: %s\n", kc.Dir)
+	return os.MkdirAll(kc.Dir, 0755)
+}
+
+// InstallCertManager installs the cert manager bundle.
+func (kc *KbContext) InstallCertManager() error {
+	if _, err := kc.Kubectl.Command("create", "namespace", "cert-manager"); err != nil {
+		return err
+	}
+	if _, err := kc.Kubectl.Command("label", "namespace", "cert-manager", "certmanager.k8s.io/disable-validation=true"); err != nil {
+		return err
+	}
+	_, err := kc.Kubectl.Apply(false, "-f", "https://github.com/jetstack/cert-manager/releases/download/v0.8.0/cert-manager.yaml")
+	return err
+}
+
+// UninstallCertManager uninstalls the cert manager bundle.
+func (kc *KbContext) UninstallCertManager() {
+	if _, err := kc.Kubectl.Delete(false, "-f", "https://github.com/jetstack/cert-manager/releases/download/v0.8.0/cert-manager.yaml"); err != nil {
+		fmt.Fprintf(GinkgoWriter, "error when running kubectl delete during cleaning up cert manager: %v\n", err)
+	}
+	if _, err := kc.Kubectl.Delete(false, "namespace", "cert-manager"); err != nil {
+		fmt.Fprintf(GinkgoWriter, "error when cleaning up the cert manager namespace: %v\n", err)
+	}
+}
+
+// KustomizeCleanup is a helper func to run kustomize build and pipe the output to kubectl delete -f -
+func (kc *KbContext) KustomizeCleanup(dir string) {
```

Comment:
We should probably call this method `cleanupManifests` or `deleteResources` or something. Kustomize is an internal detail.